File: src/utils/math_utils.py
"""3D math utilities for transformations and calculations."""

import numpy as np
from typing import Tuple, Optional


def normalize_vector(v: np.ndarray) -> np.ndarray:
    """Normalize a vector."""
    norm = np.linalg.norm(v)
    if norm == 0:
        return v
    return v / norm


def rotation_matrix_from_vectors(forward: np.ndarray, up: np.ndarray) -> np.ndarray:
    """
    Create a rotation matrix from forward and up vectors.
    
    Args:
        forward: Forward direction vector (normalized)
        up: Up direction vector (normalized)
    
    Returns:
        3x3 rotation matrix
    """
    # Ensure vectors are normalized
    forward = normalize_vector(forward)
    up = normalize_vector(up)
    
    # Calculate right vector (cross product)
    right = np.cross(forward, up)
    right = normalize_vector(right)
    
    # Recalculate up to ensure orthogonality
    up = np.cross(right, forward)
    up = normalize_vector(up)
    
    # Build rotation matrix
    rotation = np.array([
        [right[0], up[0], -forward[0]],
        [right[1], up[1], -forward[1]],
        [right[2], up[2], -forward[2]]
    ])
    
    return rotation


def create_transform_matrix(
    position: np.ndarray,
    rotation: np.ndarray,
    scale: float = 1.0
) -> np.ndarray:
    """
    Create a 4x4 transformation matrix from position, rotation, and scale.
    
    Args:
        position: 3D position (x, y, z)
        rotation: 3x3 rotation matrix
        scale: Uniform scale factor
    
    Returns:
        4x4 transformation matrix
    """
    transform = np.eye(4, dtype=np.float32)
    
    # Apply scale
    transform[:3, :3] = rotation * scale
    
    # Apply translation
    transform[:3, 3] = position
    
    return transform


def calculate_chest_rotation(
    left_shoulder: np.ndarray,
    right_shoulder: np.ndarray,
    left_hip: np.ndarray,
    right_hip: np.ndarray
) -> np.ndarray:
    """
    Calculate chest rotation from shoulder and hip landmarks.
    
    Args:
        left_shoulder: Left shoulder 3D position
        right_shoulder: Right shoulder 3D position
        left_hip: Left hip 3D position
        right_hip: Right hip 3D position
    
    Returns:
        3x3 rotation matrix
    """
    # Calculate forward direction (perpendicular to shoulder line, pointing forward)
    shoulder_vec = right_shoulder - left_shoulder
    hip_vec = right_hip - left_hip
    
    # Average of shoulder and hip vectors gives torso direction
    torso_vec = (shoulder_vec + hip_vec) / 2.0
    
    # Forward is perpendicular to torso in the horizontal plane
    forward = np.array([-torso_vec[1], torso_vec[0], 0])
    forward = normalize_vector(forward)
    
    # Up direction (towards head)
    up = np.array([0, 0, 1])  # MediaPipe uses Z-up
    
    return rotation_matrix_from_vectors(forward, up)


def calculate_chest_position(
    left_shoulder: np.ndarray,
    right_shoulder: np.ndarray,
    offset_z: float = 0.05
) -> np.ndarray:
    """
    Calculate chest center position from shoulder landmarks.
    
    Args:
        left_shoulder: Left shoulder 3D position
        right_shoulder: Right shoulder 3D position
        offset_z: Forward offset from chest surface (meters)
    
    Returns:
        3D chest center position
    """
    # Chest center is midpoint of shoulders
    chest_center = (left_shoulder + right_shoulder) / 2.0
    
    # Project forward slightly (towards camera)
    # In MediaPipe coordinates, negative Z is towards camera
    chest_center[2] -= offset_z
    
    return chest_center


def perspective_projection_matrix(
    fov: float,
    aspect: float,
    near: float,
    far: float
) -> np.ndarray:
    """
    Create a perspective projection matrix.
    
    Args:
        fov: Field of view in degrees
        aspect: Aspect ratio (width / height)
        near: Near clipping plane
        far: Far clipping plane
    
    Returns:
        4x4 projection matrix
    """
    fov_rad = np.radians(fov)
    f = 1.0 / np.tan(fov_rad / 2.0)
    
    projection = np.zeros((4, 4), dtype=np.float32)
    projection[0, 0] = f / aspect
    projection[1, 1] = f
    projection[2, 2] = (far + near) / (near - far)
    projection[2, 3] = -1.0
    projection[3, 2] = (2.0 * far * near) / (near - far)
    
    return projection


def look_at_matrix(
    eye: np.ndarray,
    target: np.ndarray,
    up: np.ndarray = np.array([0, 0, 1])
) -> np.ndarray:
    """
    Create a look-at view matrix.
    
    Args:
        eye: Camera position
        target: Target position to look at
        up: Up direction vector
    
    Returns:
        4x4 view matrix
    """
    forward = normalize_vector(target - eye)
    right = normalize_vector(np.cross(forward, up))
    up = normalize_vector(np.cross(right, forward))
    
    view = np.eye(4, dtype=np.float32)
    view[0, :3] = right
    view[1, :3] = up
    view[2, :3] = -forward
    view[:3, 3] = -np.array([
        np.dot(right, eye),
        np.dot(up, eye),
        np.dot(-forward, eye)
    ])
    
    return view

File: js/src/renderer/utils/math-utils.ts
/**
 * 3D math utilities for transformations and calculations.
 * Ported from src/utils/math_utils.py using gl-matrix
 */

import { vec3, mat3, mat4 } from 'gl-matrix';

/**
 * Normalize a vector.
 */
export function normalizeVector(v: vec3): vec3 {
  const out = vec3.create();
  vec3.normalize(out, v);
  return out;
}

/**
 * Create a rotation matrix from forward and up vectors.
 */
export function rotationMatrixFromVectors(forward: vec3, up: vec3): mat3 {
  // Ensure vectors are normalized
  const f = normalizeVector(forward);
  const u = normalizeVector(up);
  
  // Calculate right vector (cross product)
  const right = vec3.create();
  vec3.cross(right, f, u);
  vec3.normalize(right, right);
  
  // Recalculate up to ensure orthogonality
  const newUp = vec3.create();
  vec3.cross(newUp, right, f);
  vec3.normalize(newUp, newUp);
  
  // Build rotation matrix
  const rotation = mat3.create();
  rotation[0] = right[0];
  rotation[1] = right[1];
  rotation[2] = right[2];
  rotation[3] = newUp[0];
  rotation[4] = newUp[1];
  rotation[5] = newUp[2];
  rotation[6] = -f[0];
  rotation[7] = -f[1];
  rotation[8] = -f[2];
  
  return rotation;
}

/**
 * Create a 4x4 transformation matrix from position, rotation, and scale.
 */
export function createTransformMatrix(
  position: vec3,
  rotation: mat3,
  scale: number = 1.0
): mat4 {
  const transform = mat4.create();
  
  // Copy rotation matrix to upper-left 3x3
  transform[0] = rotation[0] * scale;
  transform[1] = rotation[1] * scale;
  transform[2] = rotation[2] * scale;
  transform[4] = rotation[3] * scale;
  transform[5] = rotation[4] * scale;
  transform[6] = rotation[5] * scale;
  transform[8] = rotation[6] * scale;
  transform[9] = rotation[7] * scale;
  transform[10] = rotation[8] * scale;
  
  // Apply translation
  transform[12] = position[0];
  transform[13] = position[1];
  transform[14] = position[2];
  transform[15] = 1.0;
  
  return transform;
}

/**
 * Calculate chest rotation from shoulder and hip landmarks.
 */
export function calculateChestRotation(
  leftShoulder: vec3,
  rightShoulder: vec3,
  leftHip: vec3,
  rightHip: vec3
): mat3 {
  // Calculate forward direction (perpendicular to shoulder line, pointing forward)
  const shoulderVec = vec3.create();
  vec3.subtract(shoulderVec, rightShoulder, leftShoulder);
  
  const hipVec = vec3.create();
  vec3.subtract(hipVec, rightHip, leftHip);
  
  // Average of shoulder and hip vectors gives torso direction
  const torsoVec = vec3.create();
  vec3.add(torsoVec, shoulderVec, hipVec);
  vec3.scale(torsoVec, torsoVec, 0.5);
  
  // Forward is perpendicular to torso in the horizontal plane
  const forward = vec3.fromValues(-torsoVec[1], torsoVec[0], 0);
  vec3.normalize(forward, forward);
  
  // Up direction (towards head) - MediaPipe uses Z-up
  const up = vec3.fromValues(0, 0, 1);
  
  return rotationMatrixFromVectors(forward, up);
}

/**
 * Calculate chest center position from shoulder landmarks.
 */
export function calculateChestPosition(
  leftShoulder: vec3,
  rightShoulder: vec3,
  offsetZ: number = 0.05
): vec3 {
  // Chest center is midpoint of shoulders
  const chestCenter = vec3.create();
  vec3.add(chestCenter, leftShoulder, rightShoulder);
  vec3.scale(chestCenter, chestCenter, 0.5);
  
  // Project forward slightly (towards camera)
  // In MediaPipe coordinates, negative Z is towards camera
  chestCenter[2] -= offsetZ;
  
  return chestCenter;
}

/**
 * Create a perspective projection matrix.
 */
export function perspectiveProjectionMatrix(
  fov: number,
  aspect: number,
  near: number,
  far: number
): mat4 {
  const projection = mat4.create();
  mat4.perspective(projection, (fov * Math.PI) / 180, aspect, near, far);
  return projection;
}

/**
 * Create a look-at view matrix.
 */
export function lookAtMatrix(
  eye: vec3,
  target: vec3,
  up: vec3 = vec3.fromValues(0, 0, 1)
): mat4 {
  const view = mat4.create();
  mat4.lookAt(view, eye, target, up);
  return view;
}

